qiskit/transpiler/passes/optimization/hoare_opt.py

```diff
@@ -0,0 +1,343 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" Pass for hoare logic circuit optimization. """
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.circuit import ControlledGate, Gate
+from qiskit.extensions.unitary import UnitaryGate
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from qiskit.transpiler.exceptions import TranspilerError
+from . import _gate_extension  # pylint: disable=W0611
+try:
+    from z3 import And, Or, Not, Implies, Solver, Bool, unsat
+    HAS_Z3 = True
+except ModuleNotFoundError:
+    HAS_Z3 = False
+
+
+class HoareOptimizer(TransformationPass):
+    """ This is a transpiler pass using hoare logic circuit optimization.
+        The inner workings of this are detailed in:
+        https://arxiv.org/abs/1810.00375
+    """
+    def __init__(self, size=10):
+        """
+        Args:
+            size (int): size of gate cache, in number of gates
+        """
+        super().__init__()
+        self.solver = Solver()
```

Comment:
I think you'll need to check and raise if HAS_Z3 is not set with something like:
```
if not HAS_Z3:
    raise ImportError("Must have Z3 installed to use the HoareOptimizer Pass."
                                "To install run 'pip install z3-solver'")
```
before this line. Otherwise people who haven't installed Z3 will get a NameError when trying to initialize the class

Reply:
Oh right thanks, I've moved up the dependency check from the run method to the constructor since that indeed gets executed first.